bugfix(ops): Nullptr backed ArrayBuffers create slices with UB

aapoalas opened the pull request:
To facilitate debugging of this bug and eventually fix it.

aapoalas commented:
One option to handle this issue would be to use `Option<&mut [u8]>` instead of a plain slice in ops.

andreubotella commented:
In V8, empty `ArrayBuffer`s will (always, I think) return a null pointer for the data. In Rust, zero-sized slices can't be created from a null pointer, *but they can be created from a dangling pointer*. In fact, we fixed this same bug in the `Deref` implementation of `v8::BackingStore` by using dangling pointers: https://github.com/denoland/rusty_v8/blob/main/src/array_buffer.rs#L315-L320.

I opened denoland/rusty_v8#1131 to have `ArrayBuffer::data` return an `Option<NonNull>`, like `BackingStore` does. And with it, you could do:

```rust
let store = b.data().unwrap_or_else(NonNull::dangling).cast::<u8>();
// SAFETY: rust guarantees that lifetime of slice is no longer than the call.
unsafe { ::std::slice::from_raw_parts_mut(store.as_ptr(), byte_length) }
```

aapoalas commented:
> In V8, empty `ArrayBuffer`s will (always, I think) return a null pointer for the data. In Rust, zero-sized slices can't be created from a null pointer, _but they can be created from a dangling pointer_. In fact, we fixed this same bug in the `Deref` implementation of `v8::BackingStore` by using dangling pointers: https://github.com/denoland/rusty_v8/blob/main/src/array_buffer.rs#L315-L320.
> 
> I opened [denoland/rusty_v8#1131](https://github.com/denoland/rusty_v8/pull/1131) to have `ArrayBuffer::data` return an `Option<NonNull>`, like `BackingStore::data` does. And with it, you could do:
> 
> ```rust
> let store = b.data().unwrap_or_else(NonNull::dangling).cast::<u8>();
> // SAFETY: rust guarantees that lifetime of slice is no longer than the call.
> unsafe { ::std::slice::from_raw_parts_mut(store.as_ptr(), byte_length) }
> ```

The unfortunate thing is that for FFI at least, we need the null pointer. A dangling pointer will cause the `UnsafePointer.of` static method to lie about the pointer address of a given buffer.

So, to some degree the solution cannot be quite so clear cut, we also need something done in the ops macro stage.

andreubotella commented:
I think this is now fixed with #16826 and #16814.

aapoalas commented:
> I think this is now fixed with #16826 and #16814.

Yup, I think so. I rebased the PR, dropping out the asserts... Well, this PR essentially no longer serves a purpose I guess :D
